### DISCUSSION TITLE
SQLite in memory file URI behaviour

### DISCUSSION THREAD

**Discussion by Sparrow0hawk** [sqlalchemy/sqlalchemy#11399] [Usage Questions] [answered]
When creating an engine for a SQLite in-memory database, I observe different pool allocation if I use the [filename URI](https://www.sqlite.org/inmemorydb.html) syntax `file::memory:` rather than `:memory:`.

I find that if using the `file::memory:` syntax the engine is created with a `QueuePool` rather than a `SingletonThreadPool`. Is this expected behaviour?

```python
from sqlalchemy import create_engine

URI_1 = "sqlite:///:memory:"

URI_2 = "sqlite:///file::memory:"

engine1 = create_engine(URI_1)

engine2 = create_engine(URI_2)

print(engine1.pool)
print(engine2.pool)
```
```output
$ python main.py 
<sqlalchemy.pool.impl.SingletonThreadPool object at 0x102e921b0>
<sqlalchemy.pool.impl.QueuePool object at 0x102eb7b30>
```

Using SQLAlchemy 2.0.30 on Python 3.12.3 macOS 14.4.1

**Comment by CaselIT** [comment, marked as answer]
Hi,

The filename uri syntax is not detected as a memory database, so that's not supported. Use the other memory uri syntax

**Reply by Sparrow0hawk** [reply]
OK, is it worth documenting this explicitly? As SQLAlchemy still creates an engine that can be used with an in-memory database, but there are subtle behaviour differences.

**Reply by vimota** [reply]
I wasted a ton of time on this - could you add more detail about it? It's in the SQLite docs and Sqlachemy doesn't complain when you use it so I wouldn't have expected it to not be supported.

**Reply by CaselIT** [reply]
a PR that improves those docs can certainly be accepted. The documentation is in this file https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/dialects/sqlite/pysqlite.py

**Reply by zzzeek** [reply]
Where does this `sqlite:///file::memory:` syntax come from?    We have docs for connect strings, everything supported is illustrated at https://docs.sqlalchemy.org/en/20/dialects/sqlite.html#connect-strings and the next section on URI connections at https://docs.sqlalchemy.org/en/20/dialects/sqlite.html#uri-connections .   If you follow the examples there, I dont think you'd end up wasting time on syntaxes that come from somewhere else that we dont support.

**Reply by Sparrow0hawk** [reply]
[SQLite documentation](https://sqlite.org/inmemorydb.html) describes how you can the special `:memory:` filename when using URI filenames which leads to the format `file::memory:`.

**Reply by vimota** [reply]
Thanks for the response! I'm happy to send a PR for doc improvements. So just to clarify is the `file:memdb1?mode=memory&cache=shared` mentioned in [the `In-memory Databases And Shared Cache` section](https://sqlite.org/inmemorydb.html) also not supported? 

I was specifically trying to do this, so just want to confirm:
```
If two or more distinct but shareable in-memory databases are needed in a single process, then the [mode=memory](https://sqlite.org/uri.html#coreqp) query parameter can be used with a [URI filename](https://sqlite.org/uri.html) to create a named in-memory database
```

**Reply by mrckzgl** [reply]
Same problem here. I want to create a shared memory database. `file:memdb1?mode=memory&cache=shared` seems to also create a file on disk. How to create a shared memory database?

**Reply by CaselIT** [reply]
you can create the sqlite connection yourself by providing a custom creator when you create the engine: https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.create_engine.params.creator or with an event: https://docs.sqlalchemy.org/en/20/core/engines.html#custom-dbapi-args

**Reply by mrckzgl** [reply]
I am trying to attach a shared memory database to an async sqlite connection via:
```
await conn.execute(text("ATTACH DATABASE \"file:shouldnotbeondisk?mode=memory&cache=shared\" AS mydatabase"))
```
But the file `shouldnotbeondisk?mode=memory&cache=shared` is created on disk. Is this a problem on sqlalchemy or sqlite3 level?

EDIT: Ah according to https://sqlite.org/lang_attach.html
> The filename argument can be a [URI filename](https://sqlite.org/uri.html) if URI filename processing is enabled on the database connection. The default behavior is for URI filenames to be disabled, however [...]

might be the cause? How to enable URI filenames on the database connection?

EDIT2: This https://sqlite.org/uri.html helps further:
> In order to maintain full backwards compatibility for legacy applications, the URI filename capability is disabled by default. URI filenames can be enabled or disabled using the [SQLITE_USE_URI=1](https://sqlite.org/compile.html#use_uri) or [SQLITE_USE_URI=0](https://sqlite.org/compile.html#use_uri) compile-time options. The compile-time setting for URI filenames can be changed at start-time using the [sqlite3_config](https://sqlite.org/c3ref/config.html)([SQLITE_CONFIG_URI](https://sqlite.org/c3ref/c_config_covering_index_scan.html#sqliteconfiguri),1) or [sqlite3_config](https://sqlite.org/c3ref/config.html)([SQLITE_CONFIG_URI](https://sqlite.org/c3ref/c_config_covering_index_scan.html#sqliteconfiguri),0) configuration calls. Regardless of the compile-time or start-time settings, URI filenames can be enabled for individual database connections by including the [SQLITE_OPEN_URI](https://sqlite.org/c3ref/c_open_autoproxy.html) bit in the set of bits passed as the F parameter to [sqlite3_open_v2(N,P,F,V)](https://sqlite.org/c3ref/open.html).

now I just need to figure out how to call `sqlite3_config` from within python / sqlalchemy or how to set the URI bit on `sqlite3_open_v2`. Any help apprechiated :)

**Reply by mrckzgl** [reply]
It seems, if I just add "?uri=true" to the connection string, I am able to attach shared memory databases later on as expected using an URI string:
```
engine = create_async_engine("sqlite+aiosqlite:///" + filename + "?uri=true")
```
I guess the same applies if I create the engine directly with an uri:
```
engine = create_async_engine("sqlite+aiosqlite:///file:shouldnotbeondisk?mode=memory&cache=shared&uri=true")
```
At least no file is created on disk. So I would assume this sets the SQLITE_OPEN_URI bit. This also means no custom creator or event interception is necessary (anymore) :-) This matches the sqlalchemy docs: https://docs.sqlalchemy.org/en/20/dialects/sqlite.html#uri-connections

**Reply by mrckzgl** [reply]
So I would suppose (an updated) answer is also to just add the `uri=true` parameter on creating the connection to get full URI sqlite support instead of that it is not supported.

**Reply by CaselIT** [reply]
if that's a function you can call from sql you can just execute that sql

**Reply by mrckzgl** [reply]
@CaselIT No, these are C level sqlite3 library methods. They are not executable from within sql. But as written adding the &uri=true parameter to the filename uri enables the parsing of memory and cache paramters.
